### DISCUSSION TITLE
Unable to setup basic auth when hosted on Vercel

### DISCUSSION THREAD

**Discussion by cmonkey03** [vercel/community#705] [Help] [answered]
There doesn't seem to be a method to build basic auth into my Next application. My DNS is hosted outside Vercel and points to my app in Vercel, so I can't attach a policy there. Adding a reverse proxy there is not recommended. The middleware features of NextJS also are not functional. They work locally but when deployed to Vercel they fail, so I can't require basic auth on any endpoints there. It seems like the only option is Vercel's $150/mo plan for basic auth, which is a lot of money for any independent developer to pay ($1800/year) for something that can be built in a few lines of code.

**Comment by cmonkey03** [comment]
Also, custom servers are not allowed on Vercel: https://github.com/vercel/next.js/discussions/17719

**Comment by cmonkey03** [comment]
Good news, Next 12.2 update (https://nextjs.org/blog/next-12-2) and middleware update guide (https://nextjs.org/docs/messages/middleware-upgrade-guide) does the trick. On a side note, I had to re-install https://www.npmjs.com/package/@babel/plugin-proposal-class-properties for the Next upgrade to work on my monorepo, (see: https://stackoverflow.com/questions/35517245/error-missing-class-properties-transform)

**Reply by cmonkey03** [reply]
It seems like the middleware request parameter does not provide any headers to the middleware function however, so you can't read any Authorization headers

**Comment by cmonkey03** [comment]
It seems if you want to access request headers for purposes such as basic auth, you must disable static rendering and use server-side rendering: https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props. That seems a shame that on edge functions you cannot access request headers.

**Comment by leerob** [comment, marked as answer]
Here's an example showing how to add basic auth: https://github.com/vercel/examples/tree/main/edge-functions/basic-auth-password

**Reply by cmonkey03** [reply]
Thanks! I'm struggling with this as I'm working with a monorepo and a custom document (which may impact?), but I see when I clone this locally that it works.

**Comment by holmey** [comment]
@cmonkey03 I stumbled upon this discussion: https://github.com/orgs/vercel/discussions/2920. I gave it a try and as a result, I now have a simple Basic Authentication.
